netket/nn/symmetric_linear.py

```diff
@@ -211,35 +572,216 @@ def full_bias(self, bias):
+    if isinstance(symmetries, Graph):
+        # With graph try to find point group, otherwise default to automorphisms
+        if point_group:
+            sg = symmetries.space_group(point_group)
+            if mode == "auto":
+                mode = "fft"
+        elif symmetries._point_group:
+            sg = symmetries.space_group()
+            if mode == "auto":
+                mode = "fft"
+        else:
+            sg = symmetry_info.automorphisms()
+            if mode == "auto" or mode == "fft":
+                mode = "irreps"
+            if mode == "fft":
+                raise ValueError(
+                    "When requesting 'mode=fft' a valid point group must be specified"
+                    "in order to construct the space group"
+                )
+        if mode == "fft":
+            shape = tuple(symmetries.extent)
+
+    elif isinstance(symmetries, PermutationGroup):
+        # If we get a group and default to irrep projection
+        if mode == "auto":
+            mode = "irreps"
+        sg = symmetries
+
+    elif isinstance(symmetries, Sequence):
+        if not mode == "irreps":
+            raise ValueError("Specification of symmetries incompatible with mode")
+        return DenseEquivariantIrrep(symmetries, **kwargs)
+    else:
+        if symmetries.ndim == 2 and symmetries.shape[0] == symmetries.shape[1]:
+            if mode == "irreps":
+                raise ValueError("Specification of symmetries incompatible with mode")
+            if mode == "matrix":
+                return DenseEquivariantMatrix(symmetries, **kwargs)
+            else:
+                if shape is None:
+                    raise TypeError(
+                        "When requesting `mode=fft`, the shape of the translation group must be specified. "
+                        "Either supply the `shape` keyword argument or pass a `netket.graph.Graph` object to "
+                        "the symmetries keyword argument."
+                    )
+                else:
+                    return DenseEquivariantFFT(symmetries, shape=shape, **kwargs)
+        return ValueError("Invalid Specification of Symmetries")
+
+    if mode == "fft":
+        if shape is None:
+            raise TypeError(
+                "When requesting `mode=fft`, the shape of the translation group must be specified. "
+                "Either supply the `shape` keyword argument or pass a `netket.graph.Graph` object to "
+                "the symmetries keyword argument."
+            )
+        else:
+            return DenseEquivariantFFT(HashableArray(sg.product_table), shape=shape)
+    else:
+        if mode == "irreps":
+            irreps = tuple(
+                HashableArray(irrep) for irrep in symmetries.irrep_matrices()
+            )
+            return DenseEquivariantIrrep(irreps, **kwargs)
+        else:
+            return DenseEquivariantMatrix(
+                HashableArray(symmetries.product_table), **kwargs
+            )
```

Comment:
I'm lost on this logic. 
Also you should have a catch-all case if one misspells mode to something that is not handled and throw an error.

Reply:
There's three modes for DenseEquivariant because I'm keeping around the old version. I can probably write this in a way that is easier to follow.

Reply:
this last else statement i'd rewrite it as 

```python
if mode == "fft"
   if shape
   else
    ...
elif mode == "irrep":
  DenseEquivariantIrrep
elif mode == "matrix" or mode == "auto":
 DenseEquivanrMatrix
else:
  raise ValueError(f"Unknown mode={mode}. Valid modes are 'fft', 'matrix', 'irrep' or 'auto'."

Reply:
the fewer nested ifs the better.
Your last else with an if inside was throwing me off.
But maybe i'm missing something.

Reply:
Yeah that makes sense. I've implemented something like this